File: tests/fixtures/resend/resend-webhook-no-idempotency-fixed/01-set-dedup.ts
import { Webhook } from 'svix';

const wh = new Webhook(process.env.RESEND_WEBHOOK_SECRET ?? '');
const processed = new Set<string>();

// In-memory dedup keyed on the event's email_id.
export async function POST(req: Request) {
  const payload = await req.text();
  const event = wh.verify(payload, Object.fromEntries(req.headers as any)) as {
    type: string;
    data: { email_id: string };
  };

  if (processed.has(event.data.email_id)) {
    return new Response('duplicate', { status: 200 });
  }
  processed.add(event.data.email_id);

  return new Response('ok', { status: 200 });
}

File: tests/fixtures/resend/resend-webhook-no-idempotency-fixed/01-set-dedup.py
import os
from flask import Flask, request
from svix.webhooks import Webhook

app = Flask(__name__)
wh = Webhook(os.environ["RESEND_WEBHOOK_SECRET"])
processed = set()


@app.route("/webhook", methods=["POST"])
def handle_webhook():
    payload = request.get_data(as_text=True)
    event = wh.verify(
        payload,
        {
            "svix-id": request.headers.get("svix-id"),
            "svix-timestamp": request.headers.get("svix-timestamp"),
            "svix-signature": request.headers.get("svix-signature"),
        },
    )
    email_id = event.get("data", {}).get("email_id")
    if email_id in processed:
        return {"duplicate": True}
    processed.add(email_id)
    return {"ok": True}
